File: src-tauri/src/commands/query.rs
use std::collections::HashMap;
use crate::database::QueryResult;
use crate::utils::sql_formatter::SqlFormatter;
use crate::AppState;
use tauri::State;
use tracing::{info, instrument};

pub trait ToCommandResult<T> {
    fn to_cmd_result(self) -> Result<T, String>;
}

impl<T, E: std::fmt::Display> ToCommandResult<T> for Result<T, E> {
    fn to_cmd_result(self) -> Result<T, String> {
        self.map_err(|e| e.to_string())
    }
}

/// 格式化 SQL
#[tauri::command]
#[instrument(skip(state, sql))]
pub async fn beautify_sql(
    connection_id: String,
    sql: String,
    state: State<'_, AppState>,
) -> Result<String, String> {
    let manager = &state.connection_manager;
    
    let db_type = manager
        .get_database_type(&connection_id)
        .await
        .to_cmd_result()?;
        
    Ok(SqlFormatter::beautify(&sql, &db_type))
}

/// 执行 SQL 查询 - 支持多结果集
#[tauri::command]
#[instrument(skip(state, sql))]
pub async fn execute_query(
    connection_id: String,
    sql: String,
    database: Option<String>,
    state: State<'_, AppState>,
) -> Result<Vec<QueryResult>, String> {
    let manager = &state.connection_manager;
    
    info!(sql = %sql.replace('\n', " ").trim(), "收到执行请求");
    
    manager
        .execute_query(&connection_id, &sql, database.as_deref())
        .await
        .to_cmd_result()
}

/// 获取 SQL 执行计划
#[tauri::command]
#[instrument(skip(state, sql))]
pub async fn explain_query(
    connection_id: String,
    sql: String,
    database: Option<String>,
    state: State<'_, AppState>,
) -> Result<Vec<QueryResult>, String> {
    let manager = &state.connection_manager;
    
    info!(sql = %sql.replace('\n', " ").trim(), "收到解释请求");
    
    manager
        .explain_query(&connection_id, &sql, database.as_deref())
        .await
        .to_cmd_result()
}

#[tauri::command]
pub async fn execute_query_batch(
    connection_id: String,
    sqls: Vec<String>,
    state: State<'_, AppState>,
) -> Result<Vec<QueryResult>, String> {
    let manager = &state.connection_manager;
    let mut results = Vec::new();
    for sql in sqls {
        let res_vec = manager.execute_query(&connection_id, &sql, None).await.to_cmd_result()?;
        results.extend(res_vec);
    }
    Ok(results)
}

#[tauri::command]
pub async fn alter_table_structure(
    connection_id: String,
    database: String,
    table: String,
    schema: Option<String>,
    changes: Vec<crate::database::TableChange>,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let manager = &state.connection_manager;
    manager
        .alter_table(&connection_id, &table, schema.as_deref(), Some(&database), changes)
        .await
        .to_cmd_result()
}

#[tauri::command]
pub async fn update_table_data(
    connection_id: String,
    database: String,
    table: String,
    schema: Option<String>,
    column: String,
    value: Option<String>,
    where_conditions: HashMap<String, serde_json::Value>,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let manager = &state.connection_manager;
    manager
        .update_data(&connection_id, &table, schema.as_deref(), Some(&database), &column, value, where_conditions)
        .await
        .to_cmd_result()
}

#[tauri::command]
pub async fn delete_table_data(
    connection_id: String,
    database: String,
    table: String,
    schema: Option<String>,
    where_conditions: HashMap<String, serde_json::Value>,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let manager = &state.connection_manager;
    manager
        .delete_data(&connection_id, &table, schema.as_deref(), Some(&database), where_conditions)
        .await
        .to_cmd_result()
}

#[tauri::command]
pub async fn insert_table_data(
    _connection_id: String,
    _database: String,
    _table: String,
    _data: HashMap<String, serde_json::Value>,
    _state: State<'_, AppState>,
) -> Result<(), String> {
    Ok(())
}

File: src/api/query.ts
import { invoke } from '@tauri-apps/api/core'
import type { QueryResult } from '@/types/database'

export interface PreparedSqlStatement {
  sql: string
  can_page: boolean
}

export const queryApi = {
  /**
   * 执行 SQL 查询
   */
  async executeQuery(
    connectionId: string, 
    sql: string, 
    database?: string | null,
    queryId?: number | null
  ): Promise<QueryResult[]> {
    return invoke<QueryResult[]>('execute_query', { 
      connectionId, 
      sql, 
      database: database || null,
      queryId: queryId ?? null,
    })
  },

  /**
   * 解释 SQL 查询 (EXPLAIN)
   */
  async explainQuery(
    connectionId: string, 
    sql: string, 
    database?: string | null,
    queryId?: number | null
  ): Promise<QueryResult[]> {
    return invoke<QueryResult[]>('explain_query', { 
      connectionId, 
      sql, 
      database: database || null,
      queryId: queryId ?? null,
    })
  },

  /**
   * 批量执行 SQL
   */
  async executeQueryBatch(
    connectionId: string,
    sqls: string[],
    database?: string | null,
    queryId?: number | null
  ): Promise<QueryResult[]> {
    return invoke<QueryResult[]>('execute_query_batch', {
      connectionId,
      sqls,
      database: database || null,
      queryId: queryId ?? null,
    })
  },

  /**
   * 取消正在执行的 SQL
   */
  async cancelQuery(
    connectionId: string,
    queryId: number
  ): Promise<boolean> {
    return invoke<boolean>('cancel_query', {
      connectionId,
      queryId,
    })
  },

  /**
   * 让后端按数据库方言解析 SQL 脚本
   */
  async prepareSqlScript(
    connectionId: string,
    sql: string
  ): Promise<PreparedSqlStatement[]> {
    return invoke<PreparedSqlStatement[]>('prepare_sql_script', { connectionId, sql })
  },

  /**
   * 格式化/美化 SQL
   */
  async beautifySql(connectionId: string, sql: string): Promise<string> {
    return invoke<string>('beautify_sql', { connectionId, sql })
  },

  /**
   * 更新表数据 (参数化)
   */
  async updateTableData(params: {
    connectionId: string,
    database: string,
    table: string,
    schema?: string | null,
    column: string,
    value: string | null,
    whereConditions: Record<string, any>
  }): Promise<void> {
    return invoke('update_table_data', params)
  },

  /**
   * 删除表数据 (参数化)
   */
  async deleteTableData(params: {
    connectionId: string,
    database: string,
    table: string,
    schema?: string | null,
    whereConditions: Record<string, any>
  }): Promise<void> {
    return invoke('delete_table_data', params)
  },

  /**
   * 变更表结构
   */
  async alterTableStructure(params: {
    connectionId: string,
    database: string,
    table: string,
    schema?: string | null,
    changes: Array<{ type: string, data: any }>
  }): Promise<void> {
    return invoke('alter_table_structure', params)
  }
}
